### PR TITLE
[DRAFT]`generate` API proposal

### PR DESCRIPTION
**Pull request by yidong72**
Let's review the proposed `generate` API for models that can generate text. 
Note, it is for discussion. **DON'T** approve and merge it.

### CODE REVIEW

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+
+class OutputType(TypedDict):
+    sentences: List[str]  # output sentences
+    tokens: List[List[str]]  # output sentences borken into tokens
+    logits: List[Tensor]  # logits output for all the tokens
+    token_ids: List[Tensor]  # output sentence token ids
+
+
+class TextGenerableModelInterface:
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/text_generable_model.py:23]: Class naming conventions were described in our internal dev guide but I moved it into https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md so that anyone can see it. 

Let's rename this class to something like ``TextGeneration``

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    """
+    Interface for all models that is able to generate text.
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/text_generable_model.py:25]: grammar?

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    """
+
+    def generate(self, inputs: Union[List[str], Tensor, List[dict]], sampling_params: SamplingParam) -> OutputType:
+        """
+        Public method for all models that is able to generate text.
+        @params
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/text_generable_model.py:32]: wrong docstring format, it won't be picked up by Sphynx for autodoc generation. Can you please follow this kind of format for method docstrings: https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/_modules/nemo/core/classes/modelPT.html#ModelPT.register_artifact

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            inputs: Union[List[str], Tensor, List[dict]]
+                Can be one of the 3 types: 
+                1. List of strings. Each element of the list provides input prompt. The model will apply tokenizer on it.
+                    E.g [‘sentence’, ‘sentence2’ … ]
+                2. Pytorch Tensor of shape (batch_size, seq_length).  The sequence used as a prompt for the generation or as model inputs to the encoder. It will skip the tokenization step.
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/text_generable_model.py:37]: How to pad in case 2? (with which token?)

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+                    E.g. torch.tensor([[23,5234,23,35…], [223,323,23,23232,232] …])    
+                3. List of python dict objects. Used for prompt/p-tuning inputs where a set of key-value pairs are converted into input token embeddings for the model.
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/text_generable_model.py:39]: In prompt/p-tuning case are we only going to support strings? What about tokenized and/or tensor modes?

Comment (yidong72) [reply nemo/collections/nlp/modules/common/transformer/text_generable_model.py:39]: For Prompt/p-tuning model, once the model is trained, the embeddings for the virtual tokens are fixed. We can add virtual tokens to the vocab and expand the embedding table to include those virtual token embeddings which are generated from the prompt encoder.  This is the step that is needed for preparing the model for fastertransformer so it has the best compatibility with GPT models.

Once we finished the virtual tokens vocab expansion step, we can simply use the Tensor type token id input to generate text.

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py (hunk 1 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/transformer/text_generable_model.py:23]: changed

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py (hunk 2 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/transformer/text_generable_model.py:25]: fixed

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py (hunk 3 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/transformer/text_generable_model.py:32]: fixed

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py (hunk 4 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/transformer/text_generable_model.py:37]: we assume the input tokens are already padded since it expects a single tensor.

**Comment by okuchaiev** [comment]
what about .complete method? It is now redundant and should be gone, right?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 6]
```diff
@@ -875,6 +877,10 @@ def get_parameters(self):
                 params.append(param)
         return params
 
+    def generate(self, inputs: Union[List[str], Tensor, List[dict]], sampling_params: SamplingParam) -> OutputType:
+        # no implementation
+        return super().generate(inputs, sampling_params)
+
     def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
```
Comment (okuchaiev) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:884]: I'm wondering how will we maximize the code re-use between predict/eval/test_step and .generate?
Will you be relying on ._bucketize_gpt_inference under the hood of ".generate"?

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+from builtins import NotImplementedError, bool, int, float, str
+from typing import List, TypedDict, Union
+
+from torch import Tensor
+
+
+class SamplingParam(TypedDict):
+    max_length: int  # The maximum length of the sequence to be generated.
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/text_generable_model.py:8]: should we add an option to stop if "." was generated?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 6 above)
Comment (yidong72) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:884]: I would do the opposite. The `.generate` calls `trainer.predict` as we need it to initialize the DDPPlugin. So `predict_step` and `bucket` For `eval` and `test`, it just evaluate the LM loss which is different from text generation.  
` ._bucketize_gpt_inference` will not be used as the `text_generation_util` padded the tokens into the same length.

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py (hunk 5 above)
Comment (okuchaiev) [reply nemo/collections/nlp/modules/common/transformer/text_generable_model.py:40]: @vadam5  what do you think?

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 6 above)
Comment (okuchaiev) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:884]: what if Model was created without passing the trainer instance (non-model parallel)?

Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:884]: Even non-model parallel models require Trainer.

Comment (okuchaiev) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:884]: @ericharper I think it can be None for non-model parallel stuff, isn't it https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/nlp_model.py#L48 ?

Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:884]: for NLPModels, but megatron layers require ddp

Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:884]: https://github.com/NVIDIA/NeMo/blob/c4ec081c2a6d718f8917155d55a006b5f6fcbb27/nemo/collections/nlp/modules/common/megatron/transformer.py#L689

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py (hunk 7 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/transformer/text_generable_model.py:8]: i think compute_logprobs should be replaced by generate(compute_logprobs=True) or something like that

**Comment by okuchaiev** [comment]
"[text_generable_model.py]" - typo in filename

File: nemo/collections/nlp/modules/common/transformer/text_generable_model.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    token_ids: List[Tensor]  # output sentence token ids
+
+
+class TextGeneration:
+    """
+    Interface for all text generation models.
+    """
+
+    def generate(self, inputs: Union[List[str], Tensor, List[dict]], sampling_params: SamplingParam) -> OutputType:
+        """
```
Comment (okuchaiev) [inline nemo/collections/nlp/modules/common/transformer/text_generable_model.py:30]: can we set "sampling_params=None" by default

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 9]
```diff
@@ -875,6 +877,10 @@ def get_parameters(self):
                 params.append(param)
         return params
 
+    def generate(self, inputs: Union[List[str], Tensor, List[dict]], sampling_params: SamplingParam) -> OutputType:
```
Comment (vadam5) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:880]: Could the default be set to None in the method signature here?
